Add structured dtypes to `to_numpy_array`

rossbar opened the pull request:
#5250 added generic dtype support, which means arrays can now be created with structured dtypes, opening the path for creating multi-attribute adjacency matrices, i.e. what `to_numpy_recarray` does. This PR shows how this feature could be implemented within `to_numpy_array`. It takes the same basic approach as `to_numpy_recarray`, where the attribute names are taken from the names of the dtype fields. This fully replicates the behavior of `to_numpy_recarray` and would allow it to be removed, which I think is worthwhile since recarrays are not really a first-class citizen in NumPy, just a convenience layer for accessing structured arrays.

The thing that I don't like about this approach is that it makes `to_numpy_array` less readable. I don't think there's a good way to loop through multiple attributes that wouldn't cause a performance hit for the single-attribute (e.g. `weight="weight"`) case, so it becomes necessary to have a special branch for the multiattribute processing.

Marking this as draft as it's intended as a discussion point, but still needs work re: API and way more testing.

dschult commented:
I don't have much experience with structured arrays. But from your code it looks like a dtype with multiple attributes produces an array with retrieval syntax is `A[attr][i, j]` while a dtype without names produces an array with retrieval syntax `A[i, j]`.   Note that a dtype with a single attribute would still be `A[attr][i, j]`.  

So, I think you have to expect separate branches. And it is still readable code. You could make the place where the plit occurs more obviou with a comment line:  `# Handle dtypes with attribute names (structured arrays)` or similar. But I think the proposed code is readable.

rossbar commented:
Thanks for the feedback @dschult - I went ahead and added tests and updated the docstring to reflect the change, so now this is ready for review.

I wanted to enable the new features without making any changes to the API (no changing kwarg defaults, and no new kwargs). I think what I landed on is sensible, but I'm curious what others think so I will summarize it here:

To get adjacency arrays with multiple attributes a user must supply both a structured dtype (with named fields) *and* explicitly set `weight=None`. If a user tries to create a multi-attribute adjacency array with the default `weight="weight"`, they will get an explicit exception message. In the future, if we wanted to add more sophisticated logic to figure out what to do in the case where a user specifies both `weight` and a structured dtype then we could do so, but I thought that being explicit was the safest approach to start with.

I also decided to copy the behavior of `to_numpy_recarray` and not support structured dtypes for multigraph inputs. It shouldn't be hard to add, but I'd prefer to keep things simple for now and add that as a new feature down the road.

Note that none of the new features related to structured arrays should have any backward compatibility concerns, since structured dtypes didn't work at all with the former implementation.

dschult reviewed (approved):
This looks good to me!  I agree with your choice to delay any handling of multigraph for structured arrays. Also to raise an exception with `weight is not None` at the same time that structured arrays are used.
Thanks!
